### PR TITLE
[SD-3386] Fixup datelines using macros

### PR DESCRIPTION
**Pull request by marwoodandrew**
The problem is that the datelines we receive from some of the ingest providers/parsers do not conform to what we want. 

I've implemented a couple of macros to try to sort them out, the macros can be called as part of routing or maybe when we re-factor the ingest it might be an idea to nominate a macro to call on ingest.

We can't implement this stuff in the parsers as the same parser may be used for different providers.

### CODE REVIEW

File: server/macros/dpa_derive_dateline.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        # expect the dateline in the first 5 lines, sometimes there is what appears to be a headline preceeding it.
+        for line_num in range(0, min(len(lines), 5)):
+            city, source, the_rest = lines[line_num].partition(' (dpa) - ')
+            # test if we found a candidate and ensure that the city starts the line and is not crazy long
+            if source and lines[line_num].find(city) == 0 and len(city) < 20:
+                cities = find_cities()
+                located = [c for c in cities if c['city'].lower() == city.lower()]
+                if 'dateline' not in item:
+                    item['dateline'] = {}
+                item['dateline']['located'] = located[0] if len(located) > 0 else {'city_code': city, 'city': city,
```
Comment (sivakuna-aap) [inline server/macros/dpa_derive_dateline.py]: Don't we get date from DPA?

File: server/macros/dpa_derive_dateline_test.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        item['body_html'] = 'Dubbo (dpa) - Where on earth is that'
+        dpa_derive_dateline(item)
+        self.assertEqual(item['dateline']['located']['city'], 'Dubbo')
+        self.assertEqual(item['dateline']['located']['country'], 'Australia')
+
+    def no_match_test(self):
+        item = dict()
+        item['body_html'] = 'No dateline here'
+        dpa_derive_dateline(item)
+        self.assertNotIn('located', item)
```
Comment (sivakuna-aap) [inline server/macros/dpa_derive_dateline_test.py]: This should be `self.assertNotIn('located', item['dateline'])` as located will always be not in item

File: server/macros/noise11_derive_metadata.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    appropriately
+    :param item:
+    :param kwargs:
+    :return:
+    """
+    try:
+        if 'anpa_category' not in item:
+            category_map = superdesk.get_resource_service('vocabularies').find_one(req=None, _id='categories')
+            if category_map:
+                map_entry = next((code for code in category_map['items'] if code['qcode'] == 'e' and code['is_active']),
```
Comment (sivakuna-aap) [inline server/macros/noise11_derive_metadata.py]: Looks like this is been used in multiple places. Can we add a method to `VocabulariesService` which gives a generator for active vocabularies for a given vocabulary id?

File: server/macros/noise11_derive_metadata.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                                 None)
+                item['anpa_category'] = [{'qcode': 'e', 'name': map_entry['name']}]
+
+        if 'subject' not in item:
+            qcode = '01000000'
+            item['subject'] = [{'qcode': qcode, 'name': subject_codes[qcode]}]
+
+        cities = find_cities()
+        located = [c for c in cities if c['city'].lower() == 'sydney']
```
Comment (sivakuna-aap) [inline server/macros/noise11_derive_metadata.py]: If noise11 always comes from Australia/Sydney then we can fetch cities within a country using `find_cities(country_code='AU')`

File: server/macros/reuters_derive_dateline.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    :param items:
+    :return:
+    """
+    try:
+        html = item.get('body_html')
+        if html:
+            soup = BeautifulSoup(html)
+            pars = soup.findAll('p')
+            if len(pars) >= 2:
+                if 'byline' in item and item.get('byline') in pars[0].get_text():
```
Comment (sivakuna-aap) [inline server/macros/reuters_derive_dateline.py]: can use constant for 'byline'

File: server/macros/reuters_derive_dateline.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                    first = pars[1].get_text()
+                else:
+                    first = pars[0].get_text()
+                city, source, the_rest = first.partition(' (Reuters) - ')
+                if source:
+                    # sometimes the city is followed by a comma and either a date or a state
+                    city = city.split(',')[0]
+                    if any(char.isdigit() for char in city):
+                        return
+                    cities = find_cities()
```
Comment (sivakuna-aap) [inline server/macros/reuters_derive_dateline.py]: These 2-3 lines are copied across modules. Can we extract it to a function in ingest module?

File: server/macros/reuters_derive_dateline.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+                    located = [c for c in cities if c['city'].lower() == city.lower()]
+                    # if not dateline we create one
+                    if 'dateline' not in item:
+                        item['dateline'] = {}
+                    # there is already a dateline that is not Bangalore don't do anything just return
+                    elif 'located' in item['dateline'] and 'BANGALOR' != item['dateline']['located'].get(
```
Comment (sivakuna-aap) [inline server/macros/reuters_derive_dateline.py]: 'e' is missing in 'Bangalore'. Is it done deliberately?

File: server/macros/reuters_derive_dateline.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+                            'city').upper():
+                        return
+
+                    item['dateline']['located'] = located[0] if len(located) > 0 else {'city_code': city,
+                                                                                       'city': city,
+                                                                                       'tz': 'UTC',
+                                                                                       'dateline': 'city'}
+                    item['dateline']['source'] = item.get('original_source', 'Reuters')
+                    item['dateline']['text'] = city
```
Comment (sivakuna-aap) [inline server/macros/reuters_derive_dateline.py]: I think the text should be equal to what we get from reuters.

File: server/macros/dpa_derive_dateline.py (hunk 1 above)
Comment (marwoodandrew) [inline server/macros/dpa_derive_dateline.py]: In the footer we get something like "071341 GMT Okt 15" [sic] I assume that's 15th October, will need to check the spec for the definition.

Comment (marwoodandrew) [inline server/macros/dpa_derive_dateline.py]: This is defined as the transmission date and time.

File: server/macros/reuters_derive_dateline.py (hunk 8 above)
Comment (marwoodandrew) [inline server/macros/reuters_derive_dateline.py]: If Reuters send a "dateline" it seems to actually be a  date string something like "2015-10-26 02:45:12 GMT+00:00", It is not always present, it seems to depend on the channel.

File: server/macros/noise11_derive_metadata.py (hunk 3 above)
Comment (marwoodandrew) [inline server/macros/noise11_derive_metadata.py]: Might do it in another PR as it will be a core change!

**Comment by amagdas** [comment]
:+1:

**Comment by amagdas** [comment]
@superdesk/developers Ping

**Comment by akintolga** [comment]
:+1:
